File: 双指针/15. 三数之和/三数之和.cpp
#include <vector>
#include<iostream>
#include <algorithm>
using namespace std;

class Solution {
    public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> finallyvalue;
        int length = nums.size();
        for (int i = 0; i < length-2;i++) {
            int targetvalue = -nums[i];
            int left = i + 1;
            int right = length - 1;
            if(nums[i] + nums[i+1] + nums[i+2] > 0) break;
            if(nums[i] + nums[length-1] + nums[length-2] < 0) continue;
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            while (left < right) {
                int value = nums[left] + nums[right];
                if (value < targetvalue) {
                    left++;
                } else if (value > targetvalue) {
                    right--;
                } else {
                    finallyvalue.push_back({nums[i], nums[left], nums[right]});
                    left++;
                    right--;
                    while (left < right && nums[left] == nums[left - 1]) {
                        left++;
                    }
                    while (left < right && nums[right] == nums[right + 1]) {
                        right--;
                    }
                }
            }
        }
        return finallyvalue;
        }
    };

File: 双指针/15. 三数之和/三数之和.ts
function threeSum(nums: number[]): number[][] {
    let reslut: number[][] = [];
    // 接着进行排序处理操作
    nums.sort((a, b) => a - b);
    if (nums.length < 3) return reslut;
    // 遍历数组
    for (let i = 0; i < nums.length - 2; i++) {
        const frist = nums[i];
        if (frist > 0) break;
        // 跳过重复元素
        if (i > 0 && frist === nums[i - 1]) continue;
        let left: number = i + 1;
        let right: number = nums.length - 1;
        while (left < right) {
            const sum: number = frist + nums[left] + nums[right];
            if (sum > 0) right--;
            else if (sum < 0) left++;
            else {
                reslut.push([frist, nums[left], nums[right]]);
                // 跳过重复元素
                while (left < right && nums[left] === nums[left + 1]) left++;
                while (left < right && nums[right] === nums[right - 1]) right--;
                left++;
                right--;
            }
        }
    }
    return reslut;
}
